### PR TITLE
Routing Queue Client Refactor

### PR DESCRIPTION
**Pull request by shreayp**
## Overview

Description:
Implementation of all client interfaces and data structures needed for routing queues.

Why should this be merged: 

Related issue(s) (if applicable): #<number>

## Checklist (Definition of Done):

- [ ] There are no TODOs left in the code
- [ ] [Coding conventions](https://github.com/CorfuDB/CorfuDB/wiki/Corfu-Style-Guidelines) (e.g. for logging, unit tests) have been followed
- [ ] Change is covered by automated tests
- [ ] Public API has Javadoc

### CODE REVIEW

File: runtime/src/main/java/org/corfudb/runtime/collections/DeltaMessage.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
+package org.corfudb.runtime.collections;
+
+import lombok.Getter;
+
+import javax.annotation.Nullable;
+import java.util.List;
+
+public class DeltaMessage {
```
Comment (souravgarg) [inline runtime/src/main/java/org/corfudb/runtime/collections/DeltaMessage.java:8]: Aren't we using RoutingTableEntryMsg as the message type? When would DeltaMessage be used?

**Review by souravgarg — COMMENTED** [review]
This seems like the exact classes we have at AR side. Do we need all of them?

File: runtime/src/main/java/org/corfudb/runtime/collections/DeltaMessage.java (hunk 1 above)
Comment (shreayp) [reply runtime/src/main/java/org/corfudb/runtime/collections/DeltaMessage.java:8]: Yes it should be RoutingTableEntryMsg.

**Review by pankti-m — APPROVED** [review]
Looks good for now.  More changes will be required as the interfaces and protocol with the client gets more defined.

File: runtime/src/main/java/org/corfudb/runtime/RoutingQueueSenderClient.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+     * Constructor for the log replication client for routing queues on sender.
+     *
+     * @param runtime Corfu Runtime.
+     * @param clientName String representation of the client name. This parameter is case-sensitive.
+     * @throws IllegalArgumentException If clientName is null or empty.
+     * @throws InvocationTargetException InvocationTargetException.
+     * @throws NoSuchMethodException NoSuchMethodException.
+     * @throws IllegalAccessException IllegalAccessException.
+     */
+    public RoutingQueueSenderClient(CorfuRuntime runtime, String clientName)
```
Comment (souravgarg) [inline runtime/src/main/java/org/corfudb/runtime/RoutingQueueSenderClient.java:33]: In the last review with policy folks, we proposed to receive transaction context from policy. Did you add some TODO to take care of it later?

Comment (shreayp) [reply runtime/src/main/java/org/corfudb/runtime/RoutingQueueSenderClient.java:33]: Yes, transmitDeltaMessage and transmitDeltaMessages both now take in a txnContext as a parameter.

**Review by souravgarg — APPROVED** [review]

**Review by shama358 — COMMENTED** [review]
It looks good for now. Like Pankti said, some of these could change as we integrate with the client.

I also think some of the abstract classes may not be indeed. They can just be in the concrete RoutingQueue class. But we can get to them during integration.

**Review by shama358 — APPROVED** [review]

**Pull request merged**